### PR TITLE
Fixes #39253 - Implement "Not Found" error handling for invalid Job I…

### PR DESCRIPTION
**Pull request by pondrejk**
…nvocation IDs

SAT-44544

image

### CODE REVIEW

File: webpack/JobInvocationDetail/JobInvocationEmptyState.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+          <EmptyStateIcon icon={SearchIcon} />
+          <EmptyStateHeader
+            titleText={<>{sprintf(__('Job invocation not found'))}</>}
+            headingLevel="h5"
+          />
+
+          <EmptyStateBody>
+            {sprintf(
+              __(
+                'There is no job invocation with id %s or there are access permissions needed. Please contact your administrator if this issue continues.'
```
Comment (adamruzicka) [inline webpack/JobInvocationDetail/JobInvocationEmptyState.js]: From the top of my head, I could envision four error cases that can happen:

1. The user doesn't have permission to view job invocations at all (backend should give 403 unauthorized)
2. The user doesn't have permission to view this particular job invocation (backend should give 404)
3. The user has permission to view, but the job invocation doesn't exist (backend should give 404)
4. The backend failing to process the request (think 500 ISE)

Should we distinguish 1 vs 2+3 vs 4?

With that being said, the old page showed "Not found" for all cases, except 4 so feel free treat this as optional.

Comment (adamruzicka) [inline webpack/JobInvocationDetail/JobInvocationEmptyState.js]: Since we know what the permissions required are, should we list them as we do on the unauthorized page?

**Review by adamruzicka — COMMENTED** [review]
Some nitpicks, seems to work well.

File: webpack/JobInvocationDetail/JobInvocationEmptyState.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  EmptyStateVariant,
+  EmptyStateHeader,
+  EmptyStateActions,
+  EmptyStateFooter,
+} from '@patternfly/react-core';
+import { SearchIcon } from '@patternfly/react-icons';
+
+const jobInvocationsIndexPath = '/job_invocations';
+
+const JobInvocationEmptyState = ({ jobInvocationId }) => {
```
Comment (ofedoren) [inline webpack/JobInvocationDetail/JobInvocationEmptyState.js]: AFAIR we have some patterns in core:
 - EmptyState: https://github.com/theforeman/foreman/tree/develop/webpack/assets/javascripts/react_app/components/common/EmptyState
 - PermissionDenied: https://github.com/theforeman/foreman/tree/develop/webpack/assets/javascripts/react_app/components/PermissionDenied

Should we re-use them instead of implementing a brand new specific page for it? I mean if they are not feasible, that's okay, I'm just asking if it was considered? I might lag behind our UI changes, so not sure if those aren't marked for future removal.

File: webpack/JobInvocationDetail/JobInvocationEmptyState.js (hunk 1 above)
Comment (pondrejk) [reply webpack/JobInvocationDetail/JobInvocationEmptyState.js]: Would it make sense to include the backend message to the page? It could be more informative without us needing to introduce some parsing logic, wdyt?

Comment (adamruzicka) [reply webpack/JobInvocationDetail/JobInvocationEmptyState.js]: Not sure if the backend says anything reasonable, but worth a try.

File: webpack/JobInvocationDetail/JobInvocationEmptyState.js (hunk 2 above)
Comment (pondrejk) [reply webpack/JobInvocationDetail/JobInvocationEmptyState.js]: I did consider them, but found them a bit too restrictive. Otoh, if its worth it I can also extend them based on what we agree upon here

Comment (MariaAga) [reply webpack/JobInvocationDetail/JobInvocationEmptyState.js]: What about them was too restrictive? might be worth to extend them as you said

File: webpack/JobInvocationDetail/JobInvocationEmptyState.js (hunk 1 above)
Comment (MariaAga) [reply webpack/JobInvocationDetail/JobInvocationEmptyState.js]: In some cases the backend adds a nice message like for case 3 `»Resource job_invocation not found by id '88'«`

**Comment by pondrejk** [comment]
@ofedoren @MariaAga @adamruzicka since the last time:
- I changed thing so that the EmptyStatePattern from core is used
- I added a note about needed perms
- I display also the message from server

Screenshot 2026-05-12 at 15 41 38

File: webpack/JobInvocationDetail/JobInvocationEmptyState.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  PageSectionVariants,
+} from '@patternfly/react-core';
+import { SearchIcon } from '@patternfly/react-icons';
+import { translate as __, sprintf } from 'foremanReact/common/I18n';
+import { foremanUrl } from 'foremanReact/common/helpers';
+import EmptyStatePattern from 'foremanReact/components/common/EmptyState/EmptyStatePattern';
+
+const jobInvocationsIndexPath = '/job_invocations';
+
+const JobInvocationEmptyState = ({ jobInvocationId, errorMessage = null }) => {
```
Comment (ofedoren) [inline webpack/JobInvocationDetail/JobInvocationEmptyState.js]: We already use defaultProps for that
```suggestion
const JobInvocationEmptyState = ({ jobInvocationId, errorMessage }) => {
```

**Review by ofedoren — COMMENTED** [review]
It's more a high level review, I'd rather leave the React stuff for someone else :)

1. One of the reasons why I wanted to re-use or enhance and re-use something from Foreman is because we're still trying to unify the UI across the pages at some degree.

Looking at the recent change to models page in Foreman and here I see a bunch of common things, essentially a lot of repetitions, which makes it harder to unify in the future if we'd need some changes. Anyway, at least now these two pages have at least different strings:

Image

Image

It just feels odd :/

2. We'd rather add tests, especially since we can use AI for that, just make sure it actually tests something.

3. There is yet another custom error formatter. And AFAIU there is an idea of creating a general formatter that can be re-used then in other places. Should we wait for that or at least mark somehow to revisit this one and replace after the general/common one is available.

4. Some inline comments.

**Comment by pondrejk** [comment]
> Looking at the recent change to models page in Foreman and here I see a bunch of common things, essentially a lot of repetitions, which makes it harder to unify in the future if we'd need some changes.

Well I thought the point of these patterns was to declare the overall shape of things for the implementations to fill in their own content that could be site specific. To me they look fairly unified, I can extend the foreman pattern, but then we'd need to agree on which parts should be unified and which should stay modifiable

**Comment by pondrejk** [comment]
this is now dependent on https://github.com/theforeman/foreman/pull/10986

**Review by adamruzicka — CHANGES_REQUESTED** [review]
This will also require a required foreman version bump in engine.rb, ideally as a separate commit.

**Comment by pondrejk** [comment]
> This will also require a required foreman version bump in engine.rb, ideally as a separate commit.

Hi, opened https://github.com/theforeman/foreman_remote_execution/pull/1039

File: webpack/JobInvocationDetail/JobInvocationEmptyState.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      requiredPermissions={['view_job_invocations']}
+      primaryAction={{
+        label: __('Go to job invocations'),
+        onClick: () => visit(foremanUrl(jobInvocationsIndexPath)),
+        ouiaId: 'job-invocation-empty-state-go-to-job-invocations-button',
+      }}
+      secondaryActions={[
+        {
+          label: __('Create a new job invocation'),
+          onClick: () => visit(foremanUrl('/job_invocations/new')),
```
Comment (MariaAga) [inline webpack/JobInvocationDetail/JobInvocationEmptyState.js]: It seems that `ResourceLoadFailedEmptyState` accepts urls, and uses react-router for the actions (which is faster than visit), could the `onClick`s be replaced with a url?

File: webpack/JobInvocationDetail/index.js [hunk 5]
```diff
@@ -51,9 +58,28 @@ const JobInvocationDetailPage = ({
...
-  useAPI('get', currentPermissionsUrl, {
-    key: CURRENT_PERMISSIONS,
-  });
+  const { status: permissionsApiStatus } = useAPI(
+    'get',
+    currentPermissionsUrl,
+    {
+      key: CURRENT_PERMISSIONS,
+    }
+  );
```
Comment (MariaAga) [inline webpack/JobInvocationDetail/index.js]: I do think its good to check if the permissions api failed, but I think this logic should be moved to core (in a new PR, unrelated to this one)

**Review by MariaAga — CHANGES_REQUESTED** [review]

File: webpack/JobInvocationDetail/JobInvocationEmptyState.js (hunk 4 above)
Comment (pondrejk) [reply webpack/JobInvocationDetail/JobInvocationEmptyState.js]: I tried it out but didn't work for me, maybe because /job_invocations is still a rails index page?

Comment (MariaAga) [reply webpack/JobInvocationDetail/JobInvocationEmptyState.js]: In what way did it not work? I tried it out locally and using 
```
      primaryAction={{
        label: __('Go to job invocations'),
        url: foremanUrl(jobInvocationsIndexPath),
        ouiaId: 'job-invocation-empty-state-go-to-job-invocations-button',
      }}
      secondaryActions={[
        {
          label: __('Create a new job invocation'),
          url: foremanUrl('/job_invocations/new'),
          ouiaId: 'job-invocation-empty-state-create-new-job-invocation-button',
        },
      ]}
   ```
   Works, and navigates

**Review by MariaAga — CHANGES_REQUESTED** [review]
Since the empty state unloads the job invocation page when the api is done, with an error, it causes some components (JobInvocationToolbarButtons) to unload to fast and cause console warnings. I think this could be fixed by doing something like `toolbarButtons={ jobInvocationApiStatus === API_STATUS.SUCCESS && <JobInvocationToolbarButtons jobId={id} data={items} />}`

File: webpack/JobInvocationDetail/JobInvocationEmptyState.js (hunk 4 above)
Comment (pondrejk) [reply webpack/JobInvocationDetail/JobInvocationEmptyState.js]: right, works for me to :) honestly, I don't know what was the issue last time I tried, sorry for the confusion

**Comment by pondrejk** [comment]
> Since the empty state unloads the job invocation page when the api is done, with an error, it causes some components (JobInvocationToolbarButtons) to unload to fast and cause console warnings. I think this could be fixed by doing something like `toolbarButtons={ jobInvocationApiStatus === API_STATUS.SUCCESS && <JobInvocationToolbarButtons jobId={id} data={items} />}`

thanks for noticing, I added your suggestion, but I had to also add some isMounted conditions to JobInvocationToolbarButtons for the console warning to stop

File: webpack/JobInvocationDetail/index.js (hunk 5 above)
Comment (pondrejk) [reply webpack/JobInvocationDetail/index.js]: I'm not sure how to go about this, since permissionsApiStatus is one of the things we use to find what errorMessage to send to the ResourceLoadFailedEmptyState. Can you suggest where should I put that logic in core and how I'd trigger it?

Comment (MariaAga) [reply webpack/JobInvocationDetail/index.js]: Sorry my bad, I  rechecked, and the data from `currentPermissionsUrl` is already in `useForemanPermissions` in the empty state, so no need to call it at all.

Foreman ui meta data (which goes to `useForemanPermissions`)
` permissions: (User.current.admin? ? Permission.all : User.current.permissions).pluck(:name),`
https://github.com/theforeman/foreman/blob/07ce9438d216cf7d1cef3763de66b83f22669c16/app/helpers/application_helper.rb#L422

api call to `currentPermissionsUrl` 
`@current_permissions = @user.admin? ? Permission.all : @user.permissions`
https://github.com/theforeman/foreman/blob/07ce9438d216cf7d1cef3763de66b83f22669c16/app/controllers/api/v2/permissions_controller.rb#L32

Comment (pondrejk) [reply webpack/JobInvocationDetail/index.js]: makes sense, so I removed those currentPermissionsUrl calls

**Comment by pondrejk** [comment]
@MariaAga Hi, I think I've addressed all the comments, can you plrease take a look again?

File: webpack/JobInvocationDetail/index.js [hunk 6]
```diff
@@ -123,7 +154,11 @@ const JobInvocationDetailPage = ({
       <PageLayout
         header={description}
         breadcrumbOptions={breadcrumbOptions}
-        toolbarButtons={<JobInvocationToolbarButtons jobId={id} data={items} />}
+        toolbarButtons={
+          jobInvocationApiStatus === API_STATUS.SUCCESS && (
```
Comment (MariaAga) [inline webpack/JobInvocationDetail/index.js]: The STATUS constant from foremanReact/constants has keys PENDING, RESOLVED, and ERROR — there is no SUCCESS. This comparison always evaluates to false, so
  JobInvocationToolbarButtons is never rendered regardless of API state. Should be API_STATUS.RESOLVED.

File: webpack/JobInvocationDetail/index.js [hunk 7]
```diff
@@ -88,6 +98,27 @@ const JobInvocationDetailPage = ({
...
+  const apiFailed = jobInvocationApiStatus === API_STATUS.ERROR;
+
+  const backendErrorMessage = useMemo(() => {
+    if (jobInvocationApiStatus === API_STATUS.ERROR) {
+      return jobInvocationErrorMessage || null;
+    }
+    return null;
+  }, [jobInvocationApiStatus, jobInvocationErrorMessage]);
+
+  const httpStatus = jobInvocationHttpStatus;
```
Comment (MariaAga) [inline webpack/JobInvocationDetail/index.js]: Used once, no transformation. Pass jobInvocationHttpStatus directly to the prop.

File: webpack/__mocks__/foremanReact/common/helpers.js [hunk 8]
```diff
@@ -1 +1,5 @@
 export const foremanUrl = path => `foreman${path}`;
+
+export const visit = jest.fn(url => {
+  global.window.location.href = url;
+});
```
Comment (MariaAga) [inline webpack/__mocks__/foremanReact/common/helpers.js]: not used

File: webpack/JobInvocationDetail/JobInvocationEmptyState.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import { PageSection, PageSectionVariants } from '@patternfly/react-core';
+import { translate as __ } from 'foremanReact/common/I18n';
+import { foremanUrl } from 'foremanReact/common/helpers';
+import ResourceLoadFailedEmptyState from 'foremanReact/components/common/EmptyState/ResourceLoadFailedEmptyState';
+
+const jobInvocationsIndexPath = '/job_invocations';
```
Comment (MariaAga) [inline webpack/JobInvocationDetail/JobInvocationEmptyState.js]: should live in JobInvocationConstants.js alongside jobInvocationDetailsUrl.

**Review by MariaAga — CHANGES_REQUESTED** [review]
I ran an AI review, and it caught that with the api permisisons removal, the toolbar and the actions, dont get to see the correct permissions, so it will be good to switch out the permissions checks in `webpack/JobInvocationDetail/CheckboxesActions.js` and `webpack/JobInvocationDetail/JobInvocationToolbarButtons.js` with the `usePermissions` hook. (this causes the buttons next to the search bar to always be disabled)

Also missing a test to make sure the empty state loads.

**Comment by pondrejk** [comment]
> I ran an AI review, and it caught that with the api permisisons removal, the toolbar and the actions, dont get to see the correct permissions, so it will be good to switch out the permissions checks in `webpack/JobInvocationDetail/CheckboxesActions.js` and `webpack/JobInvocationDetail/JobInvocationToolbarButtons.js` with the `usePermissions` hook. (this causes the buttons next to the search bar to always be disabled)
> 
> Also missing a test to make sure the empty state loads.

Hi @MariaAga, I did the requested updates, please take a look when you have time

File: webpack/JobInvocationDetail/__tests__/MainInformation.test.js [hunk 10]
```diff
@@ -50,6 +58,33 @@ jest.mock('foremanReact/common/hooks/API/APIHooks', () => ({
   })),
 }));
 
+jest.mock('foremanReact/common/hooks/Permissions/permissionHooks', () => ({
+  usePermissions: jest.fn(() => true),
+}));
```
Comment (MariaAga) [inline webpack/JobInvocationDetail/__tests__/MainInformation.test.js]: If the mock is the same as the mock file, no need to mock it again

File: webpack/JobInvocationDetail/__tests__/JobInvocationDetailEmptyState.test.js [hunk 11]
```diff
@@ -0,0 +1,82 @@
+import React from 'react';
+import { render, screen } from '@testing-library/react';
+import '@testing-library/jest-dom/extend-expect';
+import { createMemoryHistory } from 'history';
+import { Router } from 'react-router-dom';
+import JobInvocationEmptyState from '../JobInvocationEmptyState';
+
+jest.mock('foremanReact/common/hooks/Permissions/permissionHooks', () => ({
+  usePermissions: jest.fn(() => true),
+}));
```
Comment (MariaAga) [inline webpack/JobInvocationDetail/__tests__/JobInvocationDetailEmptyState.test.js]: If the mock is the same as the mock file, no need to mock it again

File: webpack/JobInvocationDetail/__tests__/JobInvocationDetailEmptyState.test.js [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+        </button>
+        {secondaryActions.map(action => (
+          <button key={action.ouiaId} type="button" data-testid={action.ouiaId}>
+            {action.label}
+          </button>
+        ))}
+      </div>
+    ),
+  })
+);
```
Comment (MariaAga) [inline webpack/JobInvocationDetail/__tests__/JobInvocationDetailEmptyState.test.js]: Prefer just relying on the original component, rather than mocking it.

File: webpack/JobInvocationDetail/__tests__/MainInformation.test.js [hunk 13]
```diff
@@ -50,6 +58,33 @@ jest.mock('foremanReact/common/hooks/API/APIHooks', () => ({
...
+        <p>{`The ${resourceLabel} with id ${resourceId} could not be found.`}</p>
+        {errorMessage && <p>{`Server returned: ${errorMessage}`}</p>}
+        <button type="button" data-testid={primaryAction.ouiaId}>
+          {primaryAction.label}
+        </button>
+      </div>
+    ),
+  })
+);
+
```
Comment (MariaAga) [inline webpack/JobInvocationDetail/__tests__/MainInformation.test.js]: Prefer just relying on the original component, rather than mocking it.

File: webpack/JobInvocationDetail/__tests__/MainInformation.test.js [hunk 14]
```diff
@@ -50,6 +58,33 @@ jest.mock('foremanReact/common/hooks/API/APIHooks', () => ({
   })),
 }));
```
Comment (MariaAga) [inline webpack/JobInvocationDetail/__tests__/MainInformation.test.js]: useAPI is no longer used

File: webpack/JobInvocationDetail/CheckboxesActions.js [hunk 15]
```diff
@@ -24,10 +25,7 @@ import {
   DIRECT_OPEN_HOST_LIMIT,
   templateInvocationPageUrl,
 } from './JobInvocationConstants';
-import {
-  selectHasPermission,
```
Comment (MariaAga) [inline webpack/JobInvocationDetail/CheckboxesActions.js:28]: Since this is no longer used, it can be removed from `JobInvocationSelectors`

File: webpack/JobInvocationDetail/index.js [hunk 16]
```diff
@@ -5,29 +5,33 @@ import {
...
 import JobInvocationToolbarButtons from './JobInvocationToolbarButtons';
 import { getJobInvocation, getTask } from './JobInvocationActions';
 import './JobInvocationDetail.scss';
 import {
-  CURRENT_PERMISSIONS,
   DATE_OPTIONS,
   JOB_INVOCATION_KEY,
   STATUS,
   STATUS_UPPERCASE,
-  currentPermissionsUrl,
```
Comment (MariaAga) [inline webpack/JobInvocationDetail/index.js:30]: Since this (`currentPermissionsUrl`,`CURRENT_PERMISSIONS`)  is no longer used, it can be removed from `JobInvocationConstants`

File: webpack/JobInvocationDetail/JobInvocationEmptyState.js [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+      requiredPermissions={['view_job_invocations']}
+      primaryAction={{
+        label: __('Go to job invocations'),
+        url: foremanUrl(jobInvocationsIndexPath),
+        ouiaId: 'job-invocation-empty-state-go-to-job-invocations-button',
+      }}
+      secondaryActions={[
+        {
+          label: __('Create a new job invocation'),
+          url: foremanUrl('/job_invocations/new'),
```
Comment (MariaAga) [inline webpack/JobInvocationDetail/JobInvocationEmptyState.js]: both urls should be consistent, and the  `foremanUrl` can be moved to the consts

**Review by MariaAga — CHANGES_REQUESTED** [review]
Thanks! PR logic works as expected, but needs some code changes

We prefer to test what the user sees, rather than relying on test ids, so `screen.getByTestId(
        'job-invocation-empty-state-go-to-job-invocations-button'
      )` should be changed to things like `screen.getByRole('button', { name: 'Create a new job invocation' })`

You can read more about testing guidelines here: https://github.com/theforeman/foreman/blob/develop/developer_docs/ui-testing-guidelines.asciidoc

File: webpack/JobInvocationDetail/JobInvocationSelectors.js [hunk 18]
```diff
@@ -32,13 +31,3 @@ export const selectTemplateInvocationList = state =>
 
 export const selectCurrentPermisions = state =>
   selectAPIResponse(state, CURRENT_PERMISSIONS);
```
Comment (MariaAga) [inline webpack/JobInvocationDetail/JobInvocationSelectors.js]: can also be removed since its not used, and CURRENT_PERMISSIONS is undefined.

**Comment by MariaAga** [comment]
I started testing that the page still works, but now when a job is running the buttons (create report, rerun all) are flashing, and it didnt seem to happen without the pr 

https://github.com/user-attachments/assets/18fdbc4c-48f6-4dbc-aaf5-a97262513107

**Comment by pondrejk** [comment]
> I started testing that the page still works, but now when a job is running the buttons (create report, rerun all) are flashing, and it didnt seem to happen without the pr

great chatch! I did some update to fix that with ai assistance, I'm posting it as a separate commit for reviewability

**Review by MariaAga — APPROVED** [review]
Thanks! code lgtm, and I tested it
